Add `filter_glob` and `exclude_glob` to fs.walk

atollk opened the pull request:
## Type of changes

- New feature
- Documentation / docstrings

## Checklist

- [x] I've run the latest [black](https://github.com/ambv/black) with default args on new code.
- [x] I've updated CHANGELOG.md and CONTRIBUTORS.md where appropriate.
- [x] I've added tests for new code.
- [x] I accept that @PyFilesystem/maintainers may be pedantic in the code review.

## Description

Fixes #371. Fixes #459 .
Added parameters `filter_glob` and `exclude_glob` to `fs.walk.Walker` to perform selection on the walked resources based on their global path.

althonos commented on fs/wildcard.py:
```diff
@@ -123,55 +135,92 @@ def get_matcher(patterns, case_sensitive):
 
-    There is no way to quote meta-characters.
+    def __repr__(self):
+        return "_PatternMatcherToken (" + self.token_str + ")"
```
Use a repr string here:
```python
return "_PatternMatcherToken({!r})".format(self.token_str)
```
this will make `repr` roundtrip with `eval`, which is good practice for simple objects.

atollk replied:
Thanks. TIL

althonos commented on fs/wildcard.py:
```diff
@@ -123,55 +135,92 @@ def get_matcher(patterns, case_sensitive):
+    def __init__(self, pattern, case_sensitive, accept_prefixes):
+        self.case_sensitive = case_sensitive
+        self.accept_prefixes = accept_prefixes
+
```
I'd also store the pattern for good measure as an attribute here, it could help debugging.

althonos replied:
Also, this would allow writing a roundtripping `__repr__` implementation:

```python
from .repr import make_repr

class _PatternMatcher:
    def __repr__(self):
        return make_repr(
            '_PatternMatcher', 
            self.pattern, 
            case_sensitive=(self.case_sentitive, None), 
            accept_prefixes=(self.accept_prefixes, None)
        )
```

althonos commented on fs/base.py:
```diff
@@ -1645,6 +1645,9 @@ def match(self, patterns, name):
             patterns (list, optional): A list of patterns, e.g.
                 ``['*.py']``, or `None` to match everything.
             name (str): A file or directory name (not a path)
+            accept_prefix (bool): If ``True``, the name is
```
Could you maybe add an example to the **Example** section below showing what the `accept_prefix` does? I think it'd be easier to understand that way.

althonos commented on fs/wildcard.py:
```diff
@@ -123,55 +135,92 @@ def get_matcher(patterns, case_sensitive):
+                ] == "*":
+                    asterik_len += 1
+                if asterik_len > 2:
+                    raise ValueError(
+                        pattern
+                        + " is not a valid wildcard pattern. "
+                        + "(sequence of three or more *)"
+                    )
```
Seeing these different errors, I'm tempted to add a `PatternError` to `fs.errors` here, and maybe subclass it (either `TooManyTokensError` or `UnmatchedTokenError`).

atollk replied:
I added a `PatternError` which also stores the position at which the error occurred. That seemed nicer to me than having a new class for each type of error but tell me if that isn't what you had in mind.

althonos reviewed (changes requested):
Nice job once again :wink: 

My largest concern is about whether it would be possible to remove the `_translate` function completely, and use `_PatternMatcher` in `fs.glob` as well to reduce code duplication. Otherwise, it's mostly smaller stuff.

I'll have a closer look at the actual files at some point because GitHub diffs are really not helping here, but the first read didn't raise any red flag :smile:

atollk commented on fs/glob.py:
```diff
@@ -27,6 +26,54 @@
 )  # type: LRUCache[Tuple[Text, bool], Tuple[int, bool, Pattern]]
 
 
+def _translate(pattern, case_sensitive=True):
```
Yes, pretty much, but not "moved" and rather "copied". 
I think it would be possible to just merge the two; at least, I don't see any difference in their feature set or behavior anymore. 

The only odd point I can find (from [here](https://docs.pyfilesystem.org/en/latest/globbing.html)) is:
> *.py?
> Matches all .py files and .pyi, .pyc etc in the currenct directory.

That's not the way [UNIX glob](https://man7.org/linux/man-pages/man7/glob.7.html) works (which this syntax is based on, I assume) or the way `fs.wildcard` works, nor is this actually tested anywhere. So it might just be an error in documentation.

althonos replied:
Yes, I think it's a documentation error too. The current implementation just replaces a `?` in the wildcard with a `.` in the corresponding regex.

atollk commented on fs/wildcard.py:
```diff
@@ -123,55 +135,92 @@ def get_matcher(patterns, case_sensitive):
         return lambda name: True
     if case_sensitive:
-        return partial(match_any, patterns)
+        return lambda name: match_any(patterns, name, accept_prefix)
```
Is that desired though? To my knowledge, lambdas are in general considered preferable over `functools` nowadays.

althonos replied:
I'd generally go for the opposite, for the simple reason that `lambda` functions cannot be pickled. This means there are places where you can't use them (for instance as a callback for a background thread), whereas a `partial` will work fine. 

`functools.partial` also gives you a nicer `repr` string compared to a lambda IMO.

atollk commented:
Alright, so, after taking a step back and having a look at the larger picture again, it made more sense to me to put all that logic into `fs.glob` instead of `fs.wildcard`, so I reverted the change in the latter. I also don't think the two can (or should) be merged anymore, due to the specific behavior of glob. Most notably `/**/` matches `/`, which is not what you would expect from a wildcard.

Either way, several of your previous comments became obsolete through this and I implemented the rest (like `partial` instead of `lambda`).

atollk commented:
Hi. Any update on this?

atollk commented:
I just rebased this branch onto the newest `master`, so it's ready to merge now imo.

althonos commented:
@atollk : I think you rebased against your old `master` without pulling from PyFilesystem/pyfilesystem2 first, the code still have merges from a prior PR.

atollk commented:
@althonos Should be fine now.

lurch commented on CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Added
 
+- To `fs.walk.Walker`, added parameters `filter_glob` and `exclude_glob`.
```
I think "- Added `filter_glob` and `exclude_glob` parameters to `fs.walk.Walker`." sounds more natural here.

lurch commented on fs/base.py:
```diff
@@ -1648,8 +1648,8 @@ def check(self):
 
-    def match(self, patterns, name):
-        # type: (Optional[Iterable[Text]], Text) -> bool
+    def match(self, patterns, name, accept_prefix=False):
```
Should `accept_prefix` be added to the "Arguments:" docstring here?

lurch replied:
Or was the addition of this extra parameter an error, as it doesn't actually seem to be used in this function?

lurch commented on fs/base.py:
```diff
@@ -1691,6 +1691,61 @@ def match(self, patterns, name):
+            True
+            >>> my_fs.match_glob(['*.jpg', '*.png'], '/foo.gif')
+            False
+            >>> my_fs.match_glob(['dir/file.txt'], '/dir/', accept_prefix=True)
+            True
+            >>> my_fs.match_glob(['dir/file.txt'], '/dir/gile.txt', accept_prefix=True)
+            False
```
Perhaps it'd be helpful to show an example of how accept_prefix=True/False works differently? :shrug:

lurch commented on fs/base.py:
```diff
@@ -1691,6 +1691,61 @@ def match(self, patterns, name):
+        """
+        if patterns is None:
+            return True
+        if not path or path[0] != "/":
```
I think this is the same as `if not path.startswith("/"):`

lurch commented on fs/base.py:
```diff
@@ -1691,6 +1691,61 @@ def match(self, patterns, name):
+            return True
+        if not path or path[0] != "/":
+            raise ValueError("%s needs to be a string starting with /" % path)
+        if isinstance(patterns, six.text_type):
+            raise TypeError("patterns must be a list or sequence")
```
This looks a bit weird - does it mean that passing e.g. `patterns=6` would be invalid, but not raise an error?

lurch replied:
Oh, I see this is behaviour that you copied from `match()` just above :wink:

lurch commented on fs/base.py:
```diff
@@ -1691,6 +1691,61 @@ def match(self, patterns, name):
+                ``['*.py']``, or `None` to match everything.
+            path (str): A resource path, starting with "/".
+            accept_prefix (bool): If ``True``, the path is
+                not required to match the wildcards themselves
```
Should "wildcards" here say "patterns" ?

lurch commented on fs/base.py:
```diff
@@ -1691,6 +1691,61 @@ def match(self, patterns, name):
+        names).
+
+        Arguments:
+            patterns (list, optional): A list of patterns, e.g.
+                ``['*.py']``, or `None` to match everything.
```
Looking at just the doc-strings, it's hard to see how `match_glob()` is different from `match()` ?

atollk replied:
The summaries (first sentences) differ: `Check if a name matches any of a list of wildcards.` and `Check if a path matches any of a list of glob patterns.`.

lurch commented on fs/glob.py:
```diff
@@ -27,17 +36,87 @@
 )  # type: LRUCache[Tuple[Text, bool], Tuple[int, bool, Pattern]]
 
 
+def _split_pattern_by_rec(pattern):
```
What does the `rec` in this function-name mean? I wonder if `_split_pattern_by_dir` or `_split_pattern_by_sep` would make more sense?

lurch commented on fs/glob.py:
```diff
@@ -71,6 +150,8 @@ def match(pattern, path):
     except KeyError:
         levels, recursive, re_pattern = _translate_glob(pattern, case_sensitive=True)
         _PATTERN_CACHE[(pattern, True)] = (levels, recursive, re_pattern)
+    if path and path[0] != "/":
+        path = "/" + path
```
Is this the same as https://docs.pyfilesystem.org/en/latest/reference/path.html?#fs.path.abspath ?

atollk replied:
Only with the current implementation of `abspath`, not with the general idea.

lurch commented on fs/glob.py:
```diff
@@ -91,9 +172,95 @@ def imatch(pattern, path):
+    Arguments:
+        patterns (list): A list of wildcard pattern, e.g ``["*.py",
+            "*.pyc"]``
+        name (str): A filename.
```
Says `name` here but `path` in the function definition. (Same "error" in `imatch_any` too.)

lurch commented on fs/glob.py:
```diff
@@ -27,17 +36,87 @@
+    return [pattern[i + 1 : j] for i, j in zip(indices[:-1], indices[1:])]
+
+
+def _translate(pattern, case_sensitive=True):
```
The implementation of this function looks fairly hairy - would it make sense to add some unit-tests for it?

lurch commented on fs/glob.py:
```diff
@@ -91,9 +172,95 @@ def imatch(pattern, path):
     except KeyError:
         levels, recursive, re_pattern = _translate_glob(pattern, case_sensitive=True)
```
Should this be `case_sensitive=False`? (If it does, I guess it shows how much thew unit-tests are lacking :wink: )

atollk replied:
As the function summary says: "Compare a glob pattern with a path (case sensitive)." So `True` is correct.

lurch commented on fs/glob.py:
```diff
@@ -91,9 +172,95 @@ def imatch(pattern, path):
+        case_sensitive (bool): If ``True``, then the callable will be case
+            sensitive, otherwise it will be case insensitive.
+        accept_prefix (bool): If ``True``, the name is
+            not required to match the wildcards themselves
```
wildcards -> patterns ?

lurch commented on fs/walk.py:
```diff
@@ -85,11 +87,22 @@ def __init__(
+                Resources will only be returned if their global path or
+                an extension of it matches one of the patterns.
+            exclude_glob (list, optional): If supplied, this parameter
+                should be a list of path patterns e.g. ``["foo/**/*.py"]``.
```
Not sure if it makes sense for the example patterns to be the same for both `filter_glob` and `exclude_glob`? :wink: Perhaps `["foo/**/*.pyc"]` would be a more useful example for the latter?

lurch commented on fs/glob.py:
```diff
@@ -91,9 +172,95 @@ def imatch(pattern, path):
+
+    """
+    if not patterns:
+        return lambda name: True
```
Doesn't really matter, but perhaps `name` here should be `path`? (to match up with the `path` parameter-name in `match_any` and `imatch_any`)

lurch commented on fs/glob.py:
```diff
@@ -91,9 +172,95 @@ def imatch(pattern, path):
+
+def get_matcher(patterns, case_sensitive, accept_prefix=False):
+    # type: (Iterable[Text], bool, bool) -> Callable[[Text], bool]
+    """Get a callable that matches paths against the given patterns.
```
Perhaps this should also have (a suitable variant of) the "Will return `True` if ``patterns`` is an empty list." text?

lurch commented on fs/glob.py:
```diff
@@ -91,9 +172,95 @@ def imatch(pattern, path):
+        an argument matches any of the given patterns.
+
+    Example:
+        >>> from fs import wildcard
+        >>> is_python = wildcard.get_matcher(['*.py'], True)
```
Is this supposed to be `glob` rather than `wildcard` ?

lurch commented on fs/walk.py:
```diff
@@ -85,11 +87,22 @@ def __init__(
                 any of these patterns will be removed from the walk.
             filter_dirs (list, optional): A list of patterns that will be used
                 to match directories paths. The walk will only open directories
```
Would it make sense to change "directories paths" here to "directory names"? :shrug:

lurch commented on fs/walk.py:
```diff
@@ -198,10 +215,19 @@ def _iter_walk(
     def _check_open_dir(self, fs, path, info):
         # type: (FS, Text, Info) -> bool
         """Check if a directory should be considered in the walk."""
+        full_path = ("" if path == "/" else path) + "/" + info.name
```
Is this a clunky way of doing https://docs.pyfilesystem.org/en/latest/reference/path.html#fs.path.combine ?

lurch commented on fs/glob.py:
```diff
@@ -27,17 +36,87 @@
+
+    There is no way to quote meta-characters.
+    Arguments:
+        pattern (str): A wildcard pattern.
```
A glob pattern?

wayconA commented:
Please promote this PR I'm desperately in need of that feature. 
@atollk 
@lurch 
@althonos

althonos commented:
@wayconA

Sorry for not getting this merged sooner, I've been under lots of work for my research projects. 

@atollk has definitely done a lot of work on this, and the reason I've been stalling this PR a bit is probably because I was not 100% happy with the API, but honestly I can't come up with something better, so I will merge this ASAP, and we can always refactor in the `pyfilesystem3` release (which I swear is not vaporwave).

Wallaku commented:
What needs to be done for this PR to be approved?
I'm determined to help.
@atollk 
@lurch 
@althonos

atollk commented:
done

Wallaku commented:
GREAT!!! Thank you very much @atollk 

Shall it be merged now?
@althonos @lurch

althonos commented:
Great job @atollk, and thanks for the review @lurch.

Wallaku commented:
Great! When the new release will be released? @althonos @lurch

althonos commented:
I'd just like #542 to be merged first and then we'll get a new release :+1:
